File: backend/src/ai/chunking.ts
export interface TextChunk {
  index: number;
  text: string;
  charStart: number;
  charEnd: number;
}

export const RAG_CHUNK_SIZE = 800;
export const RAG_CHUNK_OVERLAP = 200;
export const RAG_MAX_CHUNKS = 200;

export function splitSentences(normalized: string): string[] {
  return normalized
    .split(/(?<=[.!?])\s+/)
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
}

export function chunkTranscript(transcript: string, size = RAG_CHUNK_SIZE, overlap = RAG_CHUNK_OVERLAP): TextChunk[] {
  const normalized = transcript.replace(/\s+/g, ' ').trim();
  if (!normalized) return [];
  const sentences = splitSentences(normalized);
  const spans: { text: string; start: number; end: number }[] = [];
  let search = 0;
  for (const s of sentences) {
    const i = normalized.indexOf(s, search);
    const start = i >= 0 ? i : search;
    spans.push({ text: s, start, end: start + s.length });
    search = start + s.length;
  }
  const chunks: TextChunk[] = [];
  let buf: { text: string; start: number; end: number }[] = [];
  const bufText = (): string => buf.map((b) => b.text).join(' ');
  const emit = (): void => {
    if (!buf.length || chunks.length >= RAG_MAX_CHUNKS) return;
    const text = bufText();
    chunks.push({ index: chunks.length, text, charStart: buf[0].start, charEnd: buf[buf.length - 1].end });
  };
  const carryOverlap = (): void => {
    if (!buf.length) return;
    const full = bufText();
    const acc: { text: string; start: number; end: number }[] = [];
    let len = 0;
    for (let i = buf.length - 1; i >= 0; i--) {
      const add = (acc.length ? 1 : 0) + buf[i].text.length;
      if (len + add > overlap) break;
      acc.unshift(buf[i]);
      len += add;
    }
    if (!acc.length) {
      const tail = full.slice(-overlap);
      const start = buf[buf.length - 1].end - tail.length;
      buf = [{ text: tail, start, end: start + tail.length }];
      void full;
    } else {
      buf = acc;
    }
  };
  for (const s of spans) {
    let pending: { text: string; start: number; end: number } | null = s;
    while (pending) {
      const candidate = buf.length ? `${bufText()} ${pending.text}` : pending.text;
      if (candidate.length <= size) {
        buf.push(pending);
        pending = null;
      } else if (!buf.length) {
        const slice = pending.text.slice(0, size);
        chunks.push({ index: chunks.length, text: slice, charStart: pending.start, charEnd: pending.start + slice.length });
        if (chunks.length >= RAG_MAX_CHUNKS) return chunks;
        const rest = pending.text.slice(size - Math.min(overlap, size - 1));
        pending = rest ? { text: rest, start: pending.start + (pending.text.length - rest.length), end: pending.end } : null;
        void candidate;
      } else {
        emit();
        if (chunks.length >= RAG_MAX_CHUNKS) return chunks;
        carryOverlap();
      }
    }
  }
  emit();
  return chunks;
}

File: ai-service/app/rag/chunking.py
"""Transcript chunking with sentence-aware windows and character overlap.

Algorithm (documented per PRD):
1. Normalize whitespace to single spaces.
2. Split into sentences on `.`/`!`/`?` boundaries.
3. Greedily pack sentences into windows of at most CHUNK_SIZE characters.
4. Start each new window with the trailing OVERLAP characters of prior content
   so concepts spanning a boundary remain retrievable.
5. Cap at MAX_CHUNKS to bound embedding cost.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

DEFAULT_CHUNK_SIZE = 800
DEFAULT_CHUNK_OVERLAP = 200
MAX_CHUNKS = 200


@dataclass
class Chunk:
    index: int
    text: str
    char_start: int
    char_end: int


def normalize(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip()


def split_sentences(normalized: str) -> list[str]:
    return [s.strip() for s in re.split(r"(?<=[.!?])\s+", normalized) if s.strip()]


def chunk_text(text: str, size: int = DEFAULT_CHUNK_SIZE, overlap: int = DEFAULT_CHUNK_OVERLAP) -> list[Chunk]:
    normalized = normalize(text)
    if not normalized:
        return []
    sentences = split_sentences(normalized)
    chunks: list[Chunk] = []
    buf: list[str] = []
    buf_len = 0
    cursor = 0

    def emit() -> None:
        nonlocal buf, buf_len
        if not buf:
            return
        joined = " ".join(buf)
        start = normalized.find(buf[0], max(0, cursor - len(joined) - 10))
        if start < 0:
            start = cursor
        chunks.append(Chunk(index=len(chunks), text=joined, char_start=start, char_end=start + len(joined)))
        tail = joined[-overlap:] if overlap < len(joined) else joined
        buf = [tail] if tail else []
        buf_len = len(tail)

    for sentence in sentences:
        if len(sentence) > size and not buf:
            for i in range(0, len(sentence), size - overlap):
                piece = sentence[i : i + size]
                chunks.append(Chunk(index=len(chunks), text=piece, char_start=cursor + i, char_end=cursor + i + len(piece)))
                if len(chunks) >= MAX_CHUNKS:
                    return chunks
            cursor += len(sentence) + 1
            buf, buf_len = [], 0
            continue
        add = len(sentence) + (1 if buf else 0)
        if buf_len + add > size:
            emit()
            if len(chunks) >= MAX_CHUNKS:
                return chunks
        buf.append(sentence)
        buf_len += add
    emit()
    return chunks[:MAX_CHUNKS]
